### ISSUE TITLE
codecept_debug: test dies if variable doesn't exist

### ISSUE THREAD

**Issue by RichardPillay** [Codeception/Codeception#5808]
#### What is the problem?
If you try codecept_debug($var) and $var doesn't exist, the test dies with no indication of what went wrong. Given the idea of writing tests is to keep the code simple and readable, it shouldn't really be expected to check for existence beforehand, especially when the reason for the including the the codecept_debug line is probably to check if a previous step succeeded or not while developing the test.

#### What are you trying to achieve?
I'd like to add a simple "if variable exists" check before at the start of the function. Is there anyone who can point me to the code so I can see if it's within my sphere of capabilities?

**Comment by Naktibalda** [comment]
codecept_debug doesn't do much more than `echo`. Source code: https://github.com/Codeception/Codeception/blob/4.0.1/autoload.php#L46-L51
https://github.com/Codeception/Codeception/blob/4.0.1/src/Codeception/Util/Debug.php#L31-L37

I think that you issue happens not in `codecept_debug`, but at the place where it is called.
If variable does not exists, PHP emits a Notice and PHPUnit intercepts it and throws Exception.
The consequences depend on the part of test lifecycle where it happens - if Exception is thrown during execution of a test - the test fails in the same way as for any other uncaught exception, if exception is thrown in some hook - the concequences are much worse, including dying without any output.

It is your responsibility to code more carefully, but if you can't help it, put `@` before all function calls.
You can try to set error_level setting in codeception.yaml to exclude notices:
```yml
error_level: "E_ALL & ~E_STRICT & ~E_DEPRECATED & ~E_NOTICE"
```
https://codeception.com/docs/04-FunctionalTests#Error-Reporting

**Comment by RichardPillay** [comment]
Thanks for the reply. I know it just echoes the data, but that is the problem. Step back for a moment and think about what it does and what it is used for:

If I am an inexperienced programmer, trying to write a program by trial and error, I might do something like this:

```
        $modal = new SignupForm([
            'UserLogin'         => 'some_username',
            'password'          => 'some_password'
        ]);
        codecept_debug($model);
        $user = $model->signup();
        expect($user)->true();
```

The test in which this code was just died quietly with no indication of what was wrong. The codecept_debug line was added to try to debug the problem. It just died as well. It was only after a lot of time and and a lot more codecept_debug lines that it was discovered $model was empty because the SignupForm was being assigned to $modal instead. 

This is actually what codecept_debug is used for - figuring out what is wrong when something has been done incorrectly by the programmer. It would be very much nicer if rather than just dying quietly, the output log indicated that $model did not exist, in the same way as you would be told if the value is Null.

Best Regards,
Richard

**Comment by Naktibalda** [comment]
My point is that there is nothing specific to codecept_debug in this issue, the same thing would happen if you passed undefined variable to another function or method.

**Comment by RichardPillay** [comment]
> My point is that there is nothing specific to codecept_debug in this issue, the same thing would happen if you passed undefined variable to another function or method.

I understand your point, I really do. But isn't that exactly what you do when programming? You put in checks to catch the problems that may come up. And **because** this is a debugging function, it may be given an undefined variable by a user who is getting a problem that he does not realise is being caused by an undefined variable. This happens in production sometimes, where a library has changed, and the first thing you do is to put debugging statements to find out where the issue is. 
If it dies with a message saying the variable is undefined, it is a very quick debugging session. If it simply dies quietly, it becomes a long-drawn-out debugging session where you start putting in more and more codecept_debug lines to narrow down the area before you realise what the issue is.

This is exactly what happened when symfony 4.3.1 changed the order of parameters and  codeception started falling over. The problem is that because the code had been working for ages, you did not expect that the variable was not defined and when the script just died silently, you went looking elsewhere for the problem for many hours before realising where the problem really lay.

**Comment by RichardPillay** [comment]
Actually, will it make it any easier to understand where I'm coming from if I pose this question:

What will be your reaction if every time you hovered above a mis-spelled variable name in your IDE, it crashed because it had tried to give you information on the variable but hadn't checked beforehand that it was a valid variable?

**Comment by Naktibalda** [comment]
Please make a pull request.

**Comment by RichardPillay** [comment]
Now you're talking! I'll get started. Probably have some time this weekend.

**Comment by Naktibalda** [comment]
> The test in which this code was just died quietly with no indication of what was wrong.

You must be using an old version of Codeception, there is a very visible message if the test "just died".

**Comment by RichardPillay** [comment]
I don't think so. I'm using the version shipped with Yii2. Commandline version check says 4.0.1. 

Just to be clear, I'm not saying the **test** is dying, I'm saying the **code** is dying. In a suite of 5 tests, for example, I'm getting 2 completing and then everything just stops. When I say 2 are completing, I would get 2 ticks, crosses or E symbols and then everything just stops and returns to the commandline. I may get a stack trace or I may not, but even if I do, it's not deep enough to point at my code, just within codeception.

The E is an example of where things work correctly - There is something wrong with the code of the test itself, so that a valid test cannot be completed, but rather than simply crashing, it tells you there is something wrong. That is where you would start putting in codecept_debug lines to see where the problem is

**Comment by RichardPillay** [comment]
I think I owe you guys an apology - when I forked the code to prepare for a pull request, I discovered the changes I wanted to make were already there in the current master. I haven't checked history to see when the changes were made, but it's obvious now that what I have as part of the Yii2 distribution must be older. 

I also understand your disbelief at the problem I was describing, because what I discovered when I put the library under the debugger was the problem was not with the codecept_debug code, but code that was triggered as a result of using it and as I said before, the problem I was getting had already been solved by you previously.

I appreciate you guys taking the time to answer all my questions, especially when it didn't seem likely to be a real problem.

I'll close this, and start trying to figure out why I'm not getting the latest codeception code. It may be that the Yii2 composer file has intentional restrictions to fit in with other libraries.
